File: app/src/main/java/com/example/a_place/renderizado.java
package com.example.a_place;

import android.opengl.GLES20;
import android.opengl.GLSurfaceView;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import java.nio.FloatBuffer;



import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;



public class renderizado implements GLSurfaceView.Renderer {
    private FloatBuffer buffervertices;
    private FloatBuffer bufferTexturas;


    private int idTexturaOES;
    private int ShaderPrograma;
    private android.graphics.SurfaceTexture mSurfaceTexture;
    private android.view.Surface mSurface;

    private android.graphics.SurfaceTexture surfaceTexture;


    private final float[] panel = {
            -0.5f,  0.5f, 0.0f,
            -0.5f, -0.5f, 0.0f,
            0.5f, -0.5f, 0.0f,
            -0.5f,  0.5f, 0.0f,
            0.5f, -0.5f, 0.0f,
            0.5f,  0.5f, 0.0f
    };

    private final float[] paneluv = {
            0.0f, 1.0f,
            0.0f, 1.0f,
            1.0f, 1.0f,
            0.0f, 0.0f,
            1.0f, 1.0f,
            1.0f, 0.0f
    };

    public android.view.Surface obtenerSuperficie(){
        return mSurface;
    }

    private final String codigoShader =
            "attribute vec4 vPosition;" +
                    "atribute vec4 vTexCoord;"+ // aca por lo que tengo entendido tengo atribute vec4 almacena 4 cosas en una variable
                    "varying vec2 TexCoord;" +// lo mismo aqui
                    "void main() {" + // iniciamos main
                    "gl_Position = vPosition;" + // definimos la posicion
                    "TextCoord = vTexCoord"+ // definimos el stream que se enviara a el texture 2D
                    "}";

    private final String fragmentoshader =
            "#extension GL_OES_EGL_image_external : require\\n"+ // inicia la libreria creo de imagen externa
                    "precision mediump float;" +
                    "varying vec2 TexCoord;" +
                    "uniform samplerExternalOES sTexture;" + // inicia lo que al final le daremos a el panel para que se muestre como textura
                    "void main(){" +
                    "   gl_FragColor = texture2D(sTexture, TexCoord);" + // le damos la textura y el stream de datos que definimos anteriormente es para la imagen externa
                    "}";


    private boolean paneltoque = false;

    private float pantallaAncho = 1f;
    private float pantallaAlto = 1f;

    @Override
    public void onSurfaceCreated(GL10 gl, EGLConfig config) {
        GLES20.glClearColor(0.1f, 0.1f, 0.1f, 1.0f);
        GLES20.glEnable(GLES20.GL_DEPTH_TEST);

        ByteBuffer bb = ByteBuffer.allocateDirect(panel.length * 4);

        bb.order(ByteOrder.nativeOrder());

        buffervertices = bb.asFloatBuffer();
        buffervertices.put(panel);
        buffervertices.position(0);

        ByteBuffer bbTex = ByteBuffer.allocateDirect(paneluv.length * 4);
        bbTex.order(ByteOrder.nativeOrder());
        bufferTexturas = bbTex.asFloatBuffer();
        bufferTexturas.put(paneluv);
        bufferTexturas.position(0);

        int[] texturas = new int[1];
        GLES20.glGenTextures(1, texturas, 0);
        idTexturaOES = texturas[0];

        GLES20.glBindTexture(0x8D65, idTexturaOES); // 0x8D65 es aparentemente GLES11Ext.GL_TEXTURE_EXTERNAL_OES
        GLES20.glTexParameterf(0x8D65, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR);
        GLES20.glTexParameterf(0x8D65, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR);

        mSurfaceTexture = new android.graphics.SurfaceTexture(idTexturaOES);
        mSurfaceTexture.setDefaultBufferSize(1280, 720);
        mSurface = new android.view.Surface(mSurfaceTexture);

        int shader = cargashade(GLES20.GL_VERTEX_SHADER, codigoShader);
        int fragmentoshaderman = cargashade(GLES20.GL_FRAGMENT_SHADER, fragmentoshader);

        ShaderPrograma = GLES20.glCreateProgram();

        GLES20.glAttachShader(ShaderPrograma, shader);
        GLES20.glAttachShader(ShaderPrograma, fragmentoshaderman);
        GLES20.glLinkProgram(ShaderPrograma);


    }


    @Override
    public void onSurfaceChanged(GL10 gl, int width, int height) {
        GLES20.glViewport(0, 0, width, height);
    }

    @Override
    public void onDrawFrame(GL10 gl) {
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT | GLES20.GL_DEPTH_BUFFER_BIT);

        if (mSurfaceTexture != null) {
            try { mSurfaceTexture.updateTexImage(); } catch (Exception ignored) {}
        }

        GLES20.glUseProgram(ShaderPrograma);


        int posicionman = GLES20.glGetAttribLocation(ShaderPrograma, "vPosition");
        GLES20.glEnableVertexAttribArray(posicionman);
        GLES20.glVertexAttribPointer(posicionman, 3, GLES20.GL_FLOAT, false, 0, buffervertices);

        int uvHandle = GLES20.glGetAttribLocation(ShaderPrograma, "vTexCoord");
        GLES20.glEnableVertexAttribArray(uvHandle);
        GLES20.glVertexAttribPointer(uvHandle, 2, GLES20.GL_FLOAT, false, 0, bufferTexturas);

        GLES20.glActiveTexture(GLES20.GL_TEXTURE0);
        GLES20.glBindTexture(0x8D65, idTexturaOES);
        GLES20.glUniform1i(GLES20.glGetUniformLocation(ShaderPrograma, "sTexture"), 0);

        GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 6);

        // you will see -man everywhere i short manejador to "-man" in spanish wich means handle

        int colorman = GLES20.glGetUniformLocation(ShaderPrograma, "uColor");
        if(paneltoque){
            GLES20.glUniform4f(colorman, 0.0f, 1.0f, 0.0f, 1.0f);
        }else {
            GLES20.glUniform4f(colorman, 0.0f, 1.0f, 0.88f, 1.0f);
        }

        GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 6);
        GLES20.glDisableVertexAttribArray(posicionman);
        GLES20.glDisableVertexAttribArray(uvHandle);
    }

    public void actualizartoque(float x, float y, boolean toque){
        if(!paneltoque){
            paneltoque = true;
            return;
        }
        float normalglx = (x / pantallaAncho) * 2.0f - 1.0f;
        float normalgly = (y / pantallaAlto) * 2.0f - 1.0f;

        if(normalglx >= -0.5f && normalglx <= 0.5f && normalgly >= -0.5f && normalgly <= 0.5f){
            paneltoque = true;

            float webviewX = (normalglx + 0.5f);
            float webviewY = (normalgly + 0.5f);

        }
        else{
            paneltoque = false;
        }

    }

    public int cargashade(int tipo, String codigo) {
        int shader = GLES20.glCreateShader(tipo);
        GLES20.glShaderSource(shader, codigo);
        GLES20.glCompileShader(shader);
        return shader;
    }

}

File: app/src/main/java/com/example/a_place/renderizado.kt
package com.example.a_place

import android.opengl.GLES20
import android.opengl.GLSurfaceView
import java.nio.ByteBuffer
import android.content.Context
import java.nio.ByteOrder
import java.nio.FloatBuffer
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10
import android.opengl.Matrix

class renderizado(val context: Context) : GLSurfaceView.Renderer {
    private var buffervertices: FloatBuffer? = null
    private var bufferTexturas: FloatBuffer? = null
    private var ShaderPrograma = 0
    private var anchopantalla = 1920f
    private var altopantalla = 1080f

    private val modelMatrix = FloatArray(16)
    private val panel = floatArrayOf(
        -0.5f,  0.5f, 0.0f,
        -0.5f, -0.5f, 0.0f,
        0.5f, -0.5f, 0.0f,

        -0.5f,  0.5f, 0.0f,
        0.5f, -0.5f, 0.0f,
        0.5f,  0.5f, 0.0f
    )

    private val paneluv = floatArrayOf(
        0.0f, 0.0f,
        0.0f, 1.0f,
        1.0f, 1.0f,

        0.0f, 0.0f,
        1.0f, 1.0f,
        1.0f, 0.0f
    )

    // estamos usando Standard sampler2D
    private val codigoShader = """
        attribute vec4 vPosition;
        attribute vec2 vTexCoord;
        uniform mat4 uMatrix;
        varying vec2 vTexCoordOut;
        void main() {
            gl_Position = uMatrix * vPosition;
            vTexCoordOut = vTexCoord;
        }
    """.trimIndent()

    private val fragmentoshader = """
        precision mediump float;
        varying vec2 vTexCoordOut;
        uniform sampler2D sTexture;
        uniform vec4 uColor;
        void main() {
            gl_FragColor = texture2D(sTexture, vTexCoordOut) * uColor;
        }
    """.trimIndent()

    private var paneltoque = false
    private val paneles = mutableListOf<PanelWeb>()

    fun nuevopanel(url: String, x: Float, y: Float) {
        val panelnuevo = PanelWeb(context, url, 1280, 720)
        panelnuevo.xOffset = x
        panelnuevo.yOffset = y
        paneles.add(panelnuevo)
    }

    override fun onSurfaceCreated(gl: GL10?, config: EGLConfig?) {
        GLES20.glClearColor(0.1f, 0.1f, 0.1f, 1.0f)
        GLES20.glEnable(GLES20.GL_BLEND)
        GLES20.glBlendFunc(GLES20.GL_SRC_ALPHA, GLES20.GL_ONE_MINUS_SRC_ALPHA)

        buffervertices = ByteBuffer.allocateDirect(panel.size * 4)
            .order(ByteOrder.nativeOrder()).asFloatBuffer()
            .also { it.put(panel); it.position(0) }

        bufferTexturas = ByteBuffer.allocateDirect(paneluv.size * 4)
            .order(ByteOrder.nativeOrder()).asFloatBuffer()
            .also { it.put(paneluv); it.position(0) }

        val vertShader = cargashade(GLES20.GL_VERTEX_SHADER, codigoShader)
        val fragShader = cargashade(GLES20.GL_FRAGMENT_SHADER, fragmentoshader)

        ShaderPrograma = GLES20.glCreateProgram().also {
            GLES20.glAttachShader(it, vertShader)
            GLES20.glAttachShader(it, fragShader)
            GLES20.glLinkProgram(it)

            val status = IntArray(1)
            GLES20.glGetProgramiv(it, GLES20.GL_LINK_STATUS, status, 0)
            if (status[0] == 0) {
                android.util.Log.e("renderizado", "fallo algo checa linea 95 renderizado: ${GLES20.glGetProgramInfoLog(it)}")
            }
        }
    }

    override fun onSurfaceChanged(gl: GL10?, width: Int, height: Int) {
        GLES20.glViewport(0, 0, width, height)
        anchopantalla = width.toFloat()
        altopantalla = height.toFloat()
    }

    override fun onDrawFrame(gl: GL10?) {
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT or GLES20.GL_DEPTH_BUFFER_BIT)
        GLES20.glUseProgram(ShaderPrograma)

        val posicionman = GLES20.glGetAttribLocation(ShaderPrograma, "vPosition")
        val uvHandle    = GLES20.glGetAttribLocation(ShaderPrograma, "vTexCoord")
        val matrixman = GLES20.glGetUniformLocation(ShaderPrograma, "uMatrix")
        val colorman     = GLES20.glGetUniformLocation(ShaderPrograma, "uColor")
        val textureHandle = GLES20.glGetUniformLocation(ShaderPrograma, "sTexture")

        GLES20.glEnableVertexAttribArray(posicionman)
        GLES20.glVertexAttribPointer(posicionman, 3, GLES20.GL_FLOAT, false, 0, buffervertices)

        GLES20.glEnableVertexAttribArray(uvHandle)
        GLES20.glVertexAttribPointer(uvHandle, 2, GLES20.GL_FLOAT, false, 0, bufferTexturas)

        paneles.forEach { panel ->
            if (panel.textureId == -1) panel.iniciarTextura()

            panel.update()
            panel.updateTexture()

            Matrix.setIdentityM(modelMatrix, 0)
            Matrix.translateM(modelMatrix, 0, panel.xOffset, panel.yOffset, 0f)
            Matrix.rotateM(modelMatrix, 0, panel.rotation, 0f, 0f, 1.0f)
            GLES20.glUniformMatrix4fv(matrixman, 1, false, modelMatrix, 0)
            GLES20.glUniform4f(colorman, 1.0f, 1.0f, 1.0f, 1.0f)
            GLES20.glActiveTexture(GLES20.GL_TEXTURE0) // activamos la textura)
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, panel.textureId) // ponemos la textura acompañada
            GLES20.glUniform1i(textureHandle, 0)
            GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 6) // se dibuja los arreglos de los vertices osea el cuadrado o eso creo
            /*panel.update()
            panel.updateTexture()

            GLES20.glUniform2f(offsetHandle, panel.xOffset, panel.yOffset)
            GLES20.glUniform4f(colorman, 1.0f, 1.0f, 1.0f, 1.0f)

            GLES20.glActiveTexture(GLES20.GL_TEXTURE0) // activamos la textura
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, panel.textureId) // ponemos la textura acompañada
            GLES20.glUniform1i(textureHandle, 0)

            GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 6) // se dibuja los arreglos de los vertices osea el cuadrado o eso creo*/
        }

        GLES20.glDisableVertexAttribArray(posicionman)
        GLES20.glDisableVertexAttribArray(uvHandle)
    }

    fun actualizartoque(x: Float, y: Float, toque: Boolean) {
        if (!paneltoque) { paneltoque = true; return }
        val normalglx = (x / anchopantalla) * 2.0f - 1.0f
        val normalgly = (y / altopantalla) * 2.0f - 1.0f

        paneltoque = normalglx >= -0.5f && normalglx <= 0.5f &&
                normalgly >= -0.5f && normalgly <= 0.5f // si cae dentro de el panel toque!!!
    }

    fun cargashade(tipo: Int, codigo: String): Int {
        val shader = GLES20.glCreateShader(tipo)
        GLES20.glShaderSource(shader, codigo)
        GLES20.glCompileShader(shader)

        val status = IntArray(1)
        GLES20.glGetShaderiv(shader, GLES20.GL_COMPILE_STATUS, status, 0)
        if (status[0] == 0) {
            android.util.Log.e("renderizado", "fallo algo checa linea 159 renderizado: ${GLES20.glGetShaderInfoLog(shader)}")
        }
        return shader
    }
}